desktop/cmp/dash/canvas/DashCanvasModel.js

```diff
@@ -0,0 +1,315 @@
+        this.setLayout(restoreState.initialState.layout);
+        this.viewState = restoreState.initialState.viewState;
+        this.provider?.clear();
+    }
+
+    /**
+     * Adds a view to the DashCanvas
+     * @param {string} viewSpecId
+     * @param {number} x
+     * @param {number} y
+     * @param {number} [w]
+     * @param {number} [h]
+     * @param {Object} viewState
+     * @param {string} title
+     */
+    @action
+    addView(viewSpecId, {x, y, w, h, viewState, title} = {}) {
+        const viewSpec = this.getViewSpec(viewSpecId),
+            id = `${XH.genId()}_${Date.now()}`,
+            model = new DashCanvasViewModel({
+                id,
+                viewSpec,
+                viewState,
+                title,
+                containerModel: this
+            });
+
+        x = x ?? 0;
+        y = y ?? 0;
+        h = h ?? viewSpec.height;
+        w = w ?? viewSpec.width;
+
+        this.layout = [...this.layout, {i: id, x, y, h, w}];
+        this.viewModels = [...this.viewModels, model];
+    }
+
+    /**
+     * Remove a view from the DashCanvas
+     * @param {string} id - DashCanvasViewModel id to remove from the container
+     */
+    @action
+    removeView(id) {
+        this.layout = this.layout.filter(it => it.i !== id);
+
+        const viewModel = this.viewModels.find(it => it.id === id);
+        XH.safeDestroy(viewModel);
+
+        this.viewModels = this.viewModels.filter(it => it.id !== id);
+    }
+
+    /**
+     * Rename a view in the DashCanvas
+     * @param {string} id - DashCanvasViewModel id to remove from the container
+     */
+    async renameView(id) {
+        const view = this.viewModels.find(it => it.id === id),
+            allowRename = view?.viewSpec?.allowRename;
+        if (!allowRename) return;
+        const newName = await XH.prompt({
+            message: `Rename '${view.title}' to`,
+            title: 'Rename...',
+            icon: Icon.edit(),
+            input: {
+                initialValue: view.title,
+                rules: [required]
+            }
+        });
+        if (newName) view.setTitle(newName);
+    }
+
+    //------------------------
+    // Implementation
+    //------------------------
+    @action
+    setState(state) {
+        const {layout, viewState} = state;
+        this.setLayout(layout);
+        this.viewState = viewState;
+    }
+
+    @action
+    setLayout(layout) {
+        this.layout = layout;
+    }
+
+    @debounced(1000)
+    @action
+    publishState() {
+        const {state} = this;
+        this.provider?.write({state});
+    }
+
+    @computed
+    get viewState() {
+        const ret = {};
+        this.viewModels.forEach(({id, viewSpec, title, viewState}) => {
+            ret[id] = {
+                viewSpecId: viewSpec.id,
+                title,
+                viewState
+            };
+        });
+        return ret;
+    }
+
+    set viewState(viewState) {
```

Comment:
Let's change this from a property setter to a function (`setViewState(viewState)`) for symmetry with layout - not sure why I set this up this way initially

Reply:
I might have changed this, but actually caught the discrepancy too and refactored